File: mcspring-api/mcspring-base/src/main/java/in/kyle/mcspring/command/SimpleMethodInjection.java
package in.kyle.mcspring.command;

import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Component;
import org.springframework.util.ClassUtils;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.var;

/**
 * Used to inject method parameters
 * Does not support annotated parameters
 */
@Lazy
@Component
@RequiredArgsConstructor
public class SimpleMethodInjection {
    
    private final List<Resolver> resolvers;
    
    @SneakyThrows
    public Object invoke(Method method, Object object, List<Resolver> resolvers, Object... contextObjects) {
        Object[] params = getParameters(method, resolvers, contextObjects);
        method.setAccessible(true);
        if (params.length != 0) {
            return method.invoke(object, params);
        } else {
            return method.invoke(object);
        }
    }
    
    private List<Resolver> makeResolvers(Object... contextObjects) {
        return Stream.of(contextObjects)
                .filter(Objects::nonNull)
                .map(o -> (Resolver) parameter -> ClassUtils.isAssignable(parameter.getType(),
                                                                          o.getClass())
                        ? Optional.of(o)
                        : Optional.empty())
                .collect(Collectors.toList());
    }
    
    public Object[] getParameters(Method method, List<Resolver> contextResolvers, Object... contextObjects) {
        Parameter[] parameters = method.getParameters();
        Object[] params = new Object[parameters.length];
        
        List<Resolver> methodResolvers = new ArrayList<>();
        methodResolvers.addAll(contextResolvers);
        methodResolvers.addAll(makeResolvers(contextObjects));
        methodResolvers.addAll(resolvers);
        for (int i = 0; i < parameters.length; i++) {
            Parameter parameter = parameters[i];
    
            for (int j = 0; j < methodResolvers.size(); j++) {
                Resolver methodResolver = methodResolvers.get(j);
                Optional<Object> resolved = methodResolver.resolve(parameter);
                if (resolved.isPresent()) {
                    params[i] = resolved.get();
                    var removed = methodResolvers.remove(j);
                    methodResolvers.add(removed);
                    break;
                }
            }
    
            if (params[i] == null) {
                throw new RuntimeException(
                        "Unable to resolve parameter " + parameter.getType() + " for " +
                        method.getName());
            }
        }
        return params;
    }
}

File: mcspring-api/mcspring-base/src/main/java/in/kyle/mcspring/command/SimpleMethodInjection.kt
package `in`.kyle.mcspring.command

import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component
import kotlin.reflect.KFunction
import kotlin.reflect.full.isSubclassOf
import kotlin.reflect.jvm.isAccessible
import kotlin.reflect.jvm.javaType

@Lazy
@Component
class SimpleMethodInjection(
        private val parameterResolvers: List<ParameterResolver>
) {

    fun makeResolvers(contextObjects: List<Any?>): List<ParameterResolver> {
        return contextObjects.filterNotNull().map(this::makeResolverFor)
    }

    private fun makeResolverFor(contextObject: Any): ParameterResolver {
        return object : ParameterResolver {
            override fun resolve(parameter: Class<*>): Any? {
                return contextObject.takeIf { contextObject::class.isSubclassOf(parameter.kotlin) }
            }
        }
    }

    fun callWithInjection(function: KFunction<Any>, types: List<Class<*>>, contextObjects: List<Any?>): Any {
        val contextResolvers = makeResolvers(contextObjects)
        val parameters = getParameters(types, contextResolvers)
        try {
            function.isAccessible = true
        } catch (e: RuntimeException) {
            // java compat required
        }
        return function.call(*parameters)
    }

    fun getParameters(parameters: List<Class<*>>, contextParameterResolvers: List<ParameterResolver>): Array<Any> {
        val methodResolvers = contextParameterResolvers.plus(parameterResolvers).toMutableList()

        return parameters.map { parameter: Class<*> ->
            val candidates = methodResolvers.map { it.resolve(parameter) }.toMutableList()
            val firstIndex = candidates.indexOfFirst { it != null }
            require(firstIndex != -1) { "Unable to resolve parameter $parameter for func($parameters)" }
            methodResolvers.cycleElement(firstIndex)
            candidates[firstIndex]!!
        }.toTypedArray()
    }

    private fun <T> MutableList<T>.cycleElement(index: Int): T {
        val element = removeAt(index)
        add(element)
        return element
    }
}
